File: 1026-MaximumDifferenceBetweenNodeandAncestor/1026-MaximumDifferenceBetweenNodeandAncestor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, right, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxAncestorDiff = function(root) {
    const helper = (node, currMax, currMin) => {
        if (node == null) return currMax - currMin;

        currMax = Math.max(currMax, node.val);
        currMin = Math.min(currMin, node.val);

        let leftDifference = helper(node.left, currMax, currMin);
        let rightDifference = helper(node.right, currMax, currMin);

        return Math.max(leftDifference, rightDifference);
    }
    return helper(root, root.val, root.val);
};

File: 1026-MaximumDifferenceBetweenNodeandAncestor/1026-MaximumDifferenceBetweenNodeandAncestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        #case 1: none nodes
        if root == None: return 0
        
        
        def helper(node, minCurr, maxCurr):
            #case2: if leaf node
            if node == None:
                return maxCurr - minCurr
            
            #update min and maxs
            minCurr = min(minCurr, node.val)
            maxCurr = max(maxCurr, node.val)
            
            left = helper(node.left, minCurr, maxCurr)
            right = helper(node.right, minCurr, maxCurr)
            
            return max(left, right)
        
        return helper(root, root.val, root.val)
        
